backend/audit/excel.py

```diff
@@ -1,824 +1,863 @@
+        meta_mapping,
+        FORM_SECTIONS.ADDITIONAL_EINS,
+        template["title_row"],
+    )
+    return _extract_data(file, params)
+
+
+def extract_secondary_auditors(file):
+    template_definition_path = (
+        XLSX_TEMPLATE_DEFINITION_DIR / SECONDARY_AUDITORS_TEMPLATE_DEFINITION
+    )
+    template = json.loads(template_definition_path.read_text(encoding="utf-8"))
+    params = ExtractDataParams(
+        secondary_auditors_field_mapping,
+        secondary_auditors_column_mapping,
+        meta_mapping,
+        FORM_SECTIONS.SECONDARY_AUDITORS,
+        template["title_row"],
+    )
+    return _extract_data(file, params)
+
+
+def extract_notes_to_sefa(file):
+    template_definition_path = (
+        XLSX_TEMPLATE_DEFINITION_DIR / NOTES_TO_SEFA_TEMPLATE_DEFINITION
+    )
+    template = json.loads(template_definition_path.read_text(encoding="utf-8"))
+    params = ExtractDataParams(
+        notes_to_sefa_field_mapping,
+        notes_to_sefa_column_mapping,
+        meta_mapping,
+        FORM_SECTIONS.NOTES_TO_SEFA,
+        template["title_row"],
+    )
+    return _extract_data(file, params)
+
+
+def _extract_from_column_mapping(path, row_index, column_mapping, field_name=None):
+    """Extract named ranges from column mapping"""
+    for key, value in column_mapping.items():
+        if len(value) > 2 and (
+            value[0] + "." + value[1] == path
+            or (field_name and value[0] + "." + value[1] == path + "." + field_name)
+        ):
+            return key, row_index
+    return None, None
+
+
+def _extract_from_field_mapping(path, field_mapping, field_name=None):
+    """Extract named ranges from field mapping"""
+    for key, value in field_mapping.items():
+        if len(value) == 2 and (
+            value[0] == path
+            or (field_name and value[0] == ".".join([path, field_name]))
+        ):
+            return key, None
+    return None, None
+
+
+def _extract_error_details(error):
+    if not bool(error.path):
+        print("No path found in error object")
+        return None, None, None
+    row_index = next((item for item in error.path if isinstance(item, int)), None)
+    path = ".".join([item for item in error.path if not isinstance(item, int)])
+    return path, row_index
+
+
+def _extract_key_from_award_entities(path, row_index, named_ranges):
+    if path in [AWARD_ENTITY_NAME_PATH, AWARD_ENTITY_ID_PATH]:
+        key = (
+            AWARD_ENTITY_NAME_KEY
+            if path == AWARD_ENTITY_NAME_PATH
+            else AWARD_ENTITY_ID_KEY
+        )
+        named_ranges.append((key, row_index))
+        return key
+    return None
+
+
+def _extract_validation_field_name(error):
+    try:
+        # Parse the input data as JSON
+        data_dict = error.schema
+        # Check if the input data is in the format { ... 'not': {'required': ['field_name']}}
+        if "not" in data_dict and "required" in data_dict["not"]:
+            field_name = data_dict["not"]["required"][0]
```

Comment:
Are there any future cases where we'll have a list of not-required entries? It doesn't look like there are any cases where we have more than one field name in `{... 'not': {'required': [...]`, but we might want to keep this in mind just in case.

Reply:
We could be aggressive here and update the logic to account for a list of `not required` fields